Rework reserved-contributors by introducing a state machine

Doy-lee opened the pull request:
This goes ontop of: https://github.com/oxen-io/eth-sn-contracts/pull/85

I was reviewing some of reserved contributor implementation and noticed a couple of issues. There are various quirks like:

- Every function needs to handle the `finalized` and `cancelled` flags individually
- You can't set reserved contributors on deploy
- You deploy the BLS public key separate from the Ed25519 signature (but they're linked because they sign over each other)
- Updating the Ed25519 key is separate from BLS public key, but, both the BLS and Ed25519 signature sign over each other's respective key so there's this chicken and egg problem. You _have_ to update the Ed25519 public key before the BLS key since the contract verifies the BLS signature but _not the Ed25519 signature_
- The operator _has_ to contribute in order to update the fee, keys or parameters in which during that period, the public is able to contribute and block the operator from updating their registration
- Understanding control flow and the implication of `finalize` and `cancelled` on reset and reuse (cancel _doesn't_ allow reuse, finalize _does_)
- Not being able to reset the contract until it is finalized or cancelled, and if it's cancelled, then the contract is dead and a new one needs to be spawned instead of being able to be reused

And more and as a result of adding reserved contributors the rules have become even more complicated and difficult to manage. Prior to it, the implementation was workable but I think adding reserved contributors was the tipping point in complexity and I've gone ahead and reworked it to fix these quirks.

The main one being the state machine which rids of `cancelled` and `finalized` flags. The state progresses left-to-right, at any point you can reset the contract to go back to `WaitForOperatorContrib`. In the initial state you can update the node parameters of the contract as much as you want _until_ the operator contributes which locks in the node parameters thereafter opens the contract for contribution from the public.

`WaitForOperatorContrib -> OpenForPublicContrib -> WaitForFinalized -> Finalized`

Summary of fixes below.

---

- Contribution timestamps were not being reset

- Use enums to simplify the multi-contrib state machine. Before reserved contributions, the contract was managing state implicitly by checking side-effects of operations which was just barely manageable but after reviewing the code and the additional complexity of reserved contributions I've decided to use an enum which makes the permissibility of which functions are callable very explicit and removes a bunch of checks that were heuristics to detect the potential "state" that the contract is in.

A status enum also fixes some unusual quirks of the contract such as, the contract can only be reset when it's finalized (e.g. node is added to the SN rewards contract and funds are transferred), to now it can be reset at any stage, all contributors are refunded and contribution information is reset.

This also allows merging `contributeOperatorFunds` to `contributeFunds`, we just check the status enum to determine who is allowed to contribute. This makes consumer code integration much easier, they just need to check the status() to figure out who's allowed to contribute, and they only have one endpoint to call to contribute money into the contract.

- Adding verifying of the BLS signature on update of the BLS key preventing bad BLS keys being submitted to the contract.
- Don't automatically finalize a node when the collaterialisation is met. This was previously made automatic to remove the number of steps required for an operator, however, it seems reasonable that the operator should be the one to activate the node when they're ready incase they need to do some tweaks to their server, see: https://discord.com/channels/1163265397408149564/1264510395545161758/1264510395545161758
- Fix a race condition where it's possible for someone other than the operator to block the updating of the SN/BLS params by contributing to it, breaking the precondition

```
require(contributorAddresses.length == 1, "Cannot update pubkey: Other contributors have already joined.");
```

Required for updating those params.

- Remove BLSSignatureIsInit() which is made redundant by actually verifying the signature for the key
- Make APIs consistent with supplying node parameters
  - The constructor takes the BLS key and the SN params but not the signature or reserved
  - Reset takes reserved but no the BLS key or params
  - Contribute takes the signature and reserved but not key or params
This pushes complexity to the caller as they need to construct a specific sequence that to update fields that has side effects on other state in the contract.
- Make the updating of BLS keys, update all keys in the same step `updatePubkeys`. The BLS signature signs over the Ed25519 key, and the Ed25519 signature signs over the BLS public key, they are intrinsically linked- if one changes then necessarily you have to update at minimum the other key's signature.
- Allow the operator to withdraw their contribution via `withdrawContribution` which just thunks to `reset` and exits.
- Allow updating of fee without changing the keys. This allows fast-pathing of the case where the operator wants to tweak the fee (and since the fee is not signed, this is free to do so without necessitating an update of the Ed25519 pubkey which if the Ed25519 key did change then necessitates an update of the BLS signature due to those components signing over each other)
- Remove cancel. Cancel added an additional state that _all_ functions had to check and respect which complicates the state machine. It had a usecase of preventing other users from contributing to the contract or interacting with it. This is superceeded by the status enum where we now can reset the contract to a state prior to being funded by the operator. In this mode 0 interactions are permitted except by the owner thus an operator can "cancel" a contract by calling reset and leaving the contract alone.
- Expect the operator to be in the 0th reservation slot _IF_ a reservation is passed in. This unifies code paths and in turn makes reasoning about the system simpler. Reservations can now be updated whilst with-holding the initial contribution, hence fixing the need to contribute in the same step as setting the reservation which in turn allows setting the reserved contributors when deploying the contract (which was impossible before because you'd need to approve the funds to be spent on behalf of the deployer) or resetting.
- Disallow changing of fee after operator has contributed. Before you could change the fee/BLS keys/SN key after the operator has contributed but before the first public contribution to the node which had the race condition in that someone could interrupt the update. Now that we decouple updating from contribution, it's easy to allow an operator to update before they contribute.

Allowing them to update after they contribute, but before the public would require another state which I've decided not to add. It'd require the operator to submit another transaction to "open the node for public contribution". Instead I've made it so that, _when the operator contributes_ the node becomes open to public contribution and prior to that the node registration parameters can be updated.

---

This is passing all the JS unit tests and the integration test on core (requires an update to the JSON ABI it uses and tweaking the seed nodes to have an Ed25519 key) but note that the integration tests don't currently deploy multi-contrib SNs so that's more just to confirm that it doesn't break anything from the main rewards contract.

jagerman commented:
> Don't automatically finalize a node when the collaterialisation is met. This was previously made automatic to remove the number of steps required for an operator, however, it seems reasonable that the operator should be the one to activate the node when they're ready incase they need to do some tweaks to their server, see: https://discord.com/channels/1163265397408149564/1264510395545161758/1264510395545161758

I think this should be opt-in (or opt-out), but not forced.  While it can be useful in some cases, in the vast majority of the time a multi-contract will be set up when a node is fully synced and ready to go and you really want activation as soon as the contributions are full without having to take an extra step.  (This is particularly true for larger/more experienced operators who have their node setup process well-refined).  Having the admin need to do an extra step for submission is going to be a nuisance rather than a helpful feature for many operators.

But there *are* cases where it might be useful, particularly for smaller, or newer, operators and so having a persistent flag in the contract to turn on manual operator confirmation does seem useful.

jagerman commented:
Aside from my comment above about making manual finalization optional, this PR looks great.  I think the cleanup in state logic here is an excellent change.

Doy-lee commented:
> Aside from my comment above about making manual finalization optional, this PR looks great. I think the cleanup in state logic here is an excellent change.

Thanks! Also fair point, I added a flag `manualFinalize` that you can specify when you deploy the contract but also when you reset the contract. For most people this will be set to false and can exposed as a checkbox on the website. With false it will automatically call finalize on the last contributor.

If they want to manually do it, they can set it to true and then watch the contract for the `Filled` event if they want to get notified of when the staking requirement is met so that they can head over and invoke `finalize()` themselves manually.

I haven't merged this yet because of the requested changes in https://github.com/oxen-io/eth-sn-contracts/pull/85 which I have fixed, pending approval. This PR has been rebased on the fixes that were requested in #85.
